internal/policy/container/has_modified_files.go

```diff
@@ -45,118 +48,155 @@ func (p *HasModifiedFilesCheck) Validate(ctx context.Context, imgRef image.Image
-		var wg sync.WaitGroup
-		wg.Add(1)
-		go func() {
-			// For each path in the pathChan, add it to the layer's
-			// list of files.
-			layerFiles := make([]string, 0)
-			for path := range pathChan {
-				layerFiles = append(layerFiles, path)
+
+		if err := func() error {
+			r, err := layer.Uncompressed()
+			if err != nil {
+				return fmt.Errorf("unable to retrieve uncompressed layer contents: %w", err)
+			}
+			defer r.Close()
+
+			pathChan := make(chan string)
+
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				// For each path in the pathChan, add it to the layer's
+				// list of files.
+				layerFiles := make([]string, 0)
+				for path := range pathChan {
+					layerFiles = append(layerFiles, path)
+				}
+				// also add it to the overall list of files.
+				files[layerID.String()] = layerFiles
+			}()
+
+			if err := untar(fs, pathChan, r); err != nil {
+				return fmt.Errorf("failed to extract tarball: %w", err)
 			}
-			// also add it to the overall list of files.
-			files = append(files, layerFiles)
-			wg.Done()
-		}()
-		// add paths to the pathChan
-		err = untar(pathChan, r)
+			wg.Wait() // wait for file list to get appended to the files var
+			return nil
+		}(); err != nil {
+			return nil, err
+		}
+
+		pkgList, err := rpm.GetPackageList(ctx, layerDir)
+		if err != nil {
+			return nil, fmt.Errorf("could not get rpm list: %w", err)
+		}
+
+		pkgNameList := extractPackageNameVersion(pkgList)
+		layerPackages[layerDir] = pkgNameList
+
+		packageFiles, err := p.getInstalledFilesFor(pkgList)
 		if err != nil {
-			return nil, fmt.Errorf("failed to extract tarball: %w", err)
+			return nil, err
 		}
+		layerPackageFiles[layerDir] = packageFiles
 
-		wg.Wait() // wait for file list to get appended to the files var
+		layerIDs = append(layerIDs, layerDir)
 	}
 
-	files, dropped := p.dropFirstLayerIfEmpty(files)
-	if dropped {
-		// tell the user the first layer was dropped
-		diff0, _ := layers[0].DiffID()
-		diff1, _ := layers[1].DiffID()
-		logger.V(log.DBG).Info(
-			fmt.Sprintf("The first layer (%s) contained no files, so the next layer (%s) is being used as the base layer.",
-				diff0.String(),
-				diff1.String(),
-			))
+	packageFilesRef := packageFilesRef{
+		LayerFiles:        files,
+		LayerPackages:     layerPackages,
+		LayerPackageFiles: layerPackageFiles,
+		Layers:            layerIDs,
 	}
 
-	return &packageFilesRef{files, packageFiles}, nil
+	return &packageFilesRef, nil
 }
 
-// dropFirstLayerIfEmpty will evaluate the length of the first layer and will drop its entry in files if it's empty.
-// This avoids establishing a baseline of files installed in the original layer with an empty list. An example
-// of when this should occur is in cases where the image was built FROM scratch, such as ubi-micro. This is required
-// to ensure we don't attempt to validate against an empty layer.
-func (p *HasModifiedFilesCheck) dropFirstLayerIfEmpty(files [][]string) ([][]string, bool) {
```

Comment:
Do we no longer need to handle empty first layers?

Reply:
Since we are going layer by layer, it basically does it without having to explicitly remove it.